[Plugin] Vault

ivangonzalezacuna opened the pull request:
## Hey, I just made a Pull Request!

Hello, I've just created a vault plugin, as it was requested in #692 . It consists in a frontend and backend plugin. The backend plugin is responsible of creating an endpoint were the frontend is going to fetch the data, imilar to what jenkins does, for example. 

What's more, I've included a functionality to renew the token if it's required by the person who is integrating it. There's a built-in function which can be easily activated. It's described with the whole code in the `README.md` file.

Regarding the frontend plugin, it shows a table with the secrets inside a certain path, which is defined in a new annotation, and some links to edit or view them. I've done it in this way so that Backstage has just LIST permissions in vault, making it more secure, and it's up to the final user to have the correct privileges or not. Here you have an screenshot of the table:
![Screenshot from 2022-05-09 12-22-31](https://user-images.githubusercontent.com/49096359/167643998-54964e2b-ee3c-40e2-9a04-650d018ecfcd.png)

And in case the annotation is missing for that entity, then this will be shown:
![Screenshot from 2022-05-09 13-04-07](https://user-images.githubusercontent.com/49096359/167644385-90977bbc-4191-4d31-9f4e-00c37be9570f.png)

I still need to add some tests, but I wanted to create the PR so that you can check how it looks like. If you have any suggestion or things that would like to be different just say it here and I will try to make the changes as soon as I can.

#### :heavy_check_mark: Checklist

<!--- Please include the following in your Pull Request when applicable: -->

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [x] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

ivangonzalezacuna commented:
This should be ready now, so let me know what do you think about it :smile:

heyLu commented on plugins/vault-backend/config.d.ts:
```diff
@@ -0,0 +1,41 @@
+ * limitations under the License.
+ */
+
+/** Configuration for the Jira plugin */
```
```suggestion
/** Configuration for the Vault plugin */
```

heyLu commented on plugins/vault-backend/src/service/VaultBuilder.tsx:
```diff
@@ -0,0 +1,120 @@
+
+    router.get('/health', (_, response) => {
+      this.env.logger.info('PONG!');
+      response.send({ status: 'ok' });
```
Not sure how often this gets called, but we probably should not log this.

heyLu commented on plugins/vault-backend/src/service/vaultApi.tsx:
```diff
@@ -0,0 +1,127 @@
+/*
```
There's no React stuff in here, so this should just be a plain `.ts` file.

heyLu commented on plugins/vault/src/components/EntityVaultCard/EntityVaultCard.test.tsx:
```diff
@@ -0,0 +1,52 @@
+import { EntityVaultCard } from './EntityVaultCard';
+import { EntityProvider } from '@backstage/plugin-catalog-react';
+
+describe('EntityVautCard', () => {
```
```suggestion
describe('EntityVaultCard', () => {
```

heyLu commented on plugins/vault/src/components/EntityVaultCard/EntityVaultCard.tsx:
```diff
@@ -0,0 +1,76 @@
+        >
+          Read more
+        </Button>
+      </>
```
This should use `MissingAnnotationEmptyState`, unless there is a specific reason for hand-rolling our own.  See e.g. usage in the Kubernetes plugin: https://github.com/backstage/backstage/blob/master/plugins/techdocs/src/Router.tsx#L67

heyLu reviewed (changes requested):
Reviewing this because we work together and to help speed things along.

ivangonzalezacuna commented:
Just applied all the changes requested by @heyLu . I hope now it looks better than before

heyLu reviewed (commented):
Looks good with the changes. :+1:

heyLu commented:
I think this still needs to be added to the marketplace as described here: https://backstage.io/docs/plugins/add-to-marketplace

ivangonzalezacuna commented:
Added the missing api-reports and the file to include it to the plugin's marketplace

jhaals commented on plugins/vault-backend/src/service/runPeriodically.ts:
```diff
@@ -0,0 +1,54 @@
+ *                next.
+ * @returns A function that, when called, stops the invocation loop.
+ */
+export function runPeriodically(fn: () => any, delayMs: number): () => void {
```
This functionality can be replaced with functionality from the backend-tasks package by using the `createScheduledLocalTaskRunner` as you probably want the schedule to run on each instance running this backend.

jhaals reviewed (commented):
Nice work with this highly anticipated plugin! 😁 🙌

ivangonzalezacuna commented:
Added all the requested changes. Regarding the vault image, I have changed it for now, but I will contact the hashicorp team and ask them if it's okay to use it in this case

jhaals reviewed (commented):
Looks like the logo is updated for now 👍  I assume the new one is okay to use? :)

There are some typescript errors that needs fixing, you can spot them in the ci output or by running `yarn tsc`

jhaals commented:
@ivangonzalezacuna ah no worries! This is likely because there has been a release since you first opened the PR which causes the build to go out to npm to fetch the package instead of using the version from the monorepo :) 
Make sure that the version you depend on in package.json is the same version that's currently in master and things should hopefully work 🤞 

Except that nice work with the PR! 🙌

ivangonzalezacuna commented:
Done @jhaals . Your suggestion was right, locally it was already working again the tsc. I also updated a few other packages as well, just in case.

And thanks! I hope it can be merged and released in the next version of backstage, then we can integrate it in our company. :smile:

Rugvip commented on plugins/vault-backend/src/service/VaultBuilder.tsx:
```diff
@@ -0,0 +1,170 @@
+      if (typeof path !== 'string') {
+        res
+          .status(400)
+          .send('Something was unexpected about the path query string');
```
Avoid using `.send`, in this case it's best to throw an error instead. See https://github.com/backstage/backstage/blob/master/SECURITY.md#express-responses

Same thing with the health response above

ivangonzalezacuna replied:
Changed. I hope now everything follows the best practices :)

jhaals commented:
yarn.lock conflict left to fix and then we should be good to go 🤞

ivangonzalezacuna commented:
Finally solved most of the issues I had. Just the code scanning is failing. But I'm not sure how to fix this, as I really need to create the url path to fetch the data from vault using the data from the catalog-info.yaml file. I tried 3 different things but none of them worked.

ivangonzalezacuna commented:
@jhaals The pipeline is finally passing. So this is now ready to go :)

freben reviewed (approved):
Awesome! I had a couple of comments, but they can be deferred to a follow-up PR. So I say, we ship this now and iterate! Looking forward to this one growing and getting traction
